### ISSUE TITLE
Guards and Resolvers don't wait for APP_INITIALIZER, if RouterModule initialNavigation: 'enabled' is set

### ISSUE THREAD

**Issue by felixfiskare** [angular/universal#1623]
# 🐞 bug report

### Affected Package
@angular/router/RouterModule
@angular/core/APP_INITIALIZER

### Is this a regression?

Yes, this used to work all the way up until Angular 8.2.14.
This bug started to appear once I updated to Angular 9 (9.0.7)
EDIT: I updated to Angular 9.1, but the problem persists.

### Description

After Upgrading I noticed, that some parts of my application wouldn't start any longer, because the configuration they are depending on is empty.
I used to load this configuration from a config file on my server. I'm using APP_INITIALIZER for this. Upon further investigation I found out, that my guards and resolvers run before my APP_INITIALIZER promise resolves.
Finally I could pin-point this error to app-routing.module.ts. When I uncomment `initialNavigation: 'enabled'` then the guards and resolvers wait till the APP_INITIALIZER promise resolve.
Once enabled, the guards and resolvers run immediately and don't wait for APP_INITIALIZER.

```
@NgModule({
  imports: [
    RouterModule.forRoot(routes, {
    // initialNavigation: 'enabled'
})
  ],
  exports: [RouterModule]
})
export class AppRoutingModule { }
```

## 🔬 Minimal Reproduction

Open the console to see when different parts of the application get executed.
MinimalGuard and ResolverService should get executed after AppInitializerService Promise resolved, but unfortunately they get executed before.

https://stackblitz.com/github/felixfiskare/ResolverAndGuardBeforeInit

https://github.com/felixfiskare/ResolverAndGuardBeforeInit

## 🔥 Exception or Error

![guardAndResolverBeforeAppInit](https://user-images.githubusercontent.com/7290367/77671330-bd92bd00-6f87-11ea-9e6f-5b3bbfbc5c83.png)

## 🌍  Your Environment

**Angular Version:**
```
 ng version

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 9.0.7
Node: 13.11.0
OS: linux x64

Angular: 9.0.7
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.7
@angular-devkit/build-angular     0.900.7
@angular-devkit/build-optimizer   0.900.7
@angular-devkit/build-webpack     0.900.7
@angular-devkit/core              9.0.7
@angular-devkit/schematics        9.0.7
@ngtools/webpack                  9.0.7
@schematics/angular               9.0.7
@schematics/update                0.900.7
rxjs                              6.5.4
typescript                        3.7.5
webpack                           4.41.2
```

**Comment by CaerusKaru** [comment]
First phase of this is tracked in #1730, we'll add this section once it's merged.

**Comment by BruneXX** [comment]
Hi Guys, is this fix currently merged into v10.1.x ? thanks!

**Comment by AceVentura** [comment]
@BruneXX I tried with 10.1.1 and 10.1.2 and the issue still remains

**Comment by BruneXX** [comment]
Hi @AceVentura Sad to hear, I think we'll need to wait for this one.. Thanks

**Comment by nicolae-olariu** [comment]
Hi, any update of when is this going to be merged? Thanks!

**Comment by Abdulrrahman** [comment]
Hi, I'm facing the same issue here too

**Comment by nils-thomann** [comment]
Hi @CaerusKaru,
I'm facing the same issue. I need to wait for the `APP_INITIALIZER` before the initial navigation on the SSR. I've seen a few workarounds by manually call the `router.initialNavigation()` function at the end of the `APP_INITIALIZER` function, but unfortunately this doesn't work for me. Are there some updates on a potential fix? Or do you know a workaround for this issue? I'm currently using Angular 9 but an upgrade to Angular 10 or 11 would be possible. Thanks!

**Comment by AceVentura** [comment]
@nils-thomann the issue is still there (v11.1.1). Your only choice is to change your project to not be dependent on APP_INITIALIZER. That's what I've done.

**Comment by nils-thomann** [comment]
Hi @AceVentura, thanks for your response. Do you have some recommendations were to do stuff that has to be done before the initial navigation without using the `APP_INITIALIZER` provider? Or if you have a code example that you could show me it would be even better. Thanks for your help :)

**Comment by AceVentura** [comment]
@nils-thomann, I believe that really depends on your project. For example, in my project I have dynamic routes, which I only know by doing an API Call. 
So now I have a few assumptions in my routes (so, now it's not 100% dynamic, but like 80%) and I have a Resolver that makes the same call that I used to do in APP_INITIALIZER.

**Comment by davidediak** [comment]
Yep, same issue here.
I "resolved" by adding an "if isPlatformBrowser(this.platformId)" in each guard...

**Comment by Seemspyo** [comment]
`@nguniversal/common@11.2.1` still has no luck.

This is big issue for my case since I generally use `APP_INITIALIZER` for ensure auth(or user) states.

I tried wrapping all routes like

```ts
{ path: "", component: RootComponent, canActivate: [ EnsureAuthGuard ], children: [
  { path: "", component: SomeComponent, canActivate: [ DoSomethingWithAuthGuard ] } // errors or always not accessable.
] }
```

but as mentioned [angular repo](https://github.com/angular/angular/issues/27826), this is currently not working since all guard are runs in parallel. `CanLoad` and `CanActivateChild` are the same.

Since `initialNavigation: 'enabledBlocking'`(same as 'enabled' but 'enabled' deprecated since v11) is [required for server-side rendering](https://angular.io/api/router/InitialNavigation), as v11 it displays an warning when `initialNavigation` is not `enabled` or `enabledBlocking`.

So currently, only solution is resolve required states before all guards run.

```ts
constructor(
  private auth: EnsureAuthGuard
) { }

async canActivate() {
  await auth.ensureAuth();
  // ... now auth state surely resolved ....
}
```

Hope v12 can resolve this.

**Comment by nhanpdnguyen** [comment]
Hi guys, here's the workarounds I'm applying on my project (v11). It works with SRR as well.

- There will be a ReplaySubject `userInfoSubject` to inform the authentication status
- APP_INITIALIZER will call AuthenticationService to check for authentication status and emit the retrieved authentication status by `userInfoSuject.next(userInfo)`. Note that whenever the `userInfo` changes (by sign in, sign up...) we need to emit the authentication status too
- CanActivate will subscribe to this replay subject

```tsx
// app.module.ts
function initializeApp(
    authenticationService: AuthenticationService,
): () => Promise<any> | undefined {
    return () => {
        if (typeof window !== 'undefined') {
            return authenticationService.getUserInfo().toPromise();
        }
        return;
    };
}
```

```tsx
// authentication.service.ts
export class AuthenticationService {
    userInfo: UserLoginInfo | null | undefined;
    userInfoSubject = new ReplaySubject<UserLoginInfo | null>(1);

    constructor(
        @Inject(PLATFORM_ID) private platformId: Record<string, unknown>,
        @Optional()
        @Inject('authenticatedUserData')
        public authenticatedUserDataFromSsr: UserLoginInfo | null | undefined,
        private httpClient: HttpClient,
        private transferStateService: TransferStateService,
    ) {
        if (isPlatformServer(this.platformId)) {
            this.transferStateService.set<UserLoginInfo | null | undefined>(
                'authenticatedUserData',
                authenticatedUserDataFromSsr
            );
        }
        this.userInfo = this.transferStateService.get('authenticatedUserData');
        if (this.isFetched(this.userInfo)) {
            this.userInfoSubject.next(this.userInfo);
        }
    }

    isFetched(data: any): boolean {
        return typeof data !== 'undefined';
    }

    getUserInfo(): Observable<UserLoginInfo | null | undefined> {
        if (this.isFetched(this.userInfo)) {
            return of(this.userInfo);
        }
        return this.httpClient.get<{ user: UserLoginInfo | null }>('/userData').pipe(
            map((response) => response.user),
            catchError((err) => {
                this.userInfoSubject.next(null);
                return of(null);
            }),
            tap((userInfo) => {
                this.userInfo = userInfo;
                this.userInfoSubject.next(userInfo);
            })
        );
    }
...
}
```

```tsx
// authentication.guard.ts
export class AuthenticationGuard implements CanActivate {
    constructor(private authenticationService: AuthenticationService, private router: Router) {}

    canActivate(
        route: ActivatedRouteSnapshot,
        state: RouterStateSnapshot
    ): Observable<boolean | UrlTree> | Promise<boolean | UrlTree> | boolean | UrlTree {
        return this.authenticationService.userInfoSubject.pipe(
            map((userInfo) => {
                if (userInfo) {
                    return true;
                }
                return this.router.parseUrl('/');
            })
        );
    }
}
```

**Comment by alan-agius4** [comment]
As per above comments, this is working as intended.
